pkg/blockio/blockio.go

```diff
@@ -0,0 +1,420 @@
+// tBlockDeviceInfo.DevNode contains resolved device node, like "/dev/sda".
+type tBlockDeviceInfo struct {
+	Major   int64
+	Minor   int64
+	DevNode string
+	Origin  string
+}
+
+// Our logger instance.
+var log grclog.Logger = grclog.NewLoggerWrapper(stdlog.New(os.Stderr, "[ blockio ] ", 0))
+
+// staticOciBlockIO connects user-defined block I/O classes to
+// corresponding OCI BlockIO parameters. "Static" means that
+// new/current block devices matching device wildcards in these
+// classes are not expanded every time new containers are assigned to
+// these classes. Devices are scanned on only at the beginning and on
+// blockio configuration changes.
+var staticOciBlockIO = map[string]cgroups.OciBlockIOParameters{}
+
+// SetLogger sets the logger instance to be used by the package.
+func SetLogger(l grclog.Logger) {
+	log = l
+}
+
+// SetConfigFromFile reads and applies blockio configuration from the
+// filesystem.
+func SetConfigFromFile(filename string, force bool) error {
+	if data, err := ioutil.ReadFile(filename); err == nil {
+		if err = SetConfigFromData(data, force); err != nil {
+			return fmt.Errorf("failed to set configuration from file %q: %s", filename, err)
+		}
+		return nil
+	} else {
+		return fmt.Errorf("failed to read config file %q: %v", filename, err)
+	}
+}
+
+// SetConfigFromData parses and applies configuration from data.
+func SetConfigFromData(data []byte, force bool) error {
+	config := &Config{}
+	if err := yaml.Unmarshal(data, &config); err != nil {
+		return err
+	}
+	return SetConfig(config, force)
+}
+
+// SetConfig scans available block devices and applies new configuration
+func SetConfig(opt *Config, force bool) error {
+	if opt == nil {
+		// Setting nil configuration clears current configuration.
+		// SetConfigFromData([]byte(""), dontcare) arrives here.
+		staticOciBlockIO = map[string]cgroups.OciBlockIOParameters{}
+		return nil
+	}
+
+	currentIOSchedulers, ioSchedulerDetectionError := getCurrentIOSchedulers()
+	if ioSchedulerDetectionError != nil {
+		log.Warn("configuration validation partly disabled due to IO scheduler detection error %#v", ioSchedulerDetectionError.Error())
+	}
+
+	staticOciBlockIO = map[string]cgroups.OciBlockIOParameters{}
+	// Create static OCI BlockIO structures for each blockio class
+	for class := range opt.Classes {
+		ociBlockIO, err := devicesParametersToOci(opt.Classes[class], currentIOSchedulers)
+		if err != nil {
+			if force {
+				log.Error("ignoring: %v", err)
+			} else {
+				return err
+			}
+		}
+		// Handle all configurations as static for now. That
+		// is, the list of block devices matching Devices
+		// wildcards will not be updated without new
+		// configNotify(). class.DynamicDevices not supported
+		// yet.
+		staticOciBlockIO[class] = ociBlockIO
+	}
+	return nil
+}
+
+// GetClasses returns block I/O class names
+func GetClasses() []string {
+	classNames := make([]string, len(staticOciBlockIO))
+	i := 0
+	for name := range staticOciBlockIO {
+		i++
```

Comment:
Shouldn't you do `i++` only after the assignment?

A way to avoid the counter is to create the slice with length 0
```
classNames := make([]string, 0, len(staticOciBlockIO))
```
and use append

Reply:
Well spotted, thanks! Gee... as a penance I added a unit test even for this simple function.